File: Data Structures & Algorithms/reorder-linked-list/submission-4.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        fast = slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        
        new_start = slow.next
        slow.next = None

        prev, curr = None, new_start
        while curr:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp
        
        curr1, curr2 = head, prev
        while curr2:
            temp1 = curr1.next
            temp2 = curr2.next
            curr1.next = curr2
            curr2.next = temp1
            curr1 = temp1
            curr2 = temp2
        

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-4.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* l1 = list1;
        ListNode* l2 = list2;
        ListNode new_list = ListNode();
        ListNode* l3 = &new_list;

        while (l1 && l2) {
            if (l1->val < l2->val){
                l3->next = l1;
                l3 = l1;
                l1 = l1->next;
            }
            else {
                l3->next = l2;
                l3 = l2;
                l2 = l2 ->next;
            }
        }

        if (l1) {
            l3->next = l1;
        }
        else {
            l3->next = l2;
        }

        return new_list.next;
    }
};
